### DISCUSSION TITLE
How to validate a task form fields before is saved.

### DISCUSSION THREAD

**Discussion by IRodriguezSC** [sadieblue/agilityblue.workspace-scripts#28] [Q&A] [answered]
I have a script triggered by On Update/Before Save to validate that the entered values in the form fields meet some criteria.  The problem I'm having is that the field values on the incoming task are not coming through with the values entered.  I can do this validation on "After Safe" but the idea is to check On Before Save in order to show specific errors.

For example, when reading a quantity field on "Before Safe", the retrieved value is null:

**Read Field:**
if ($field.FieldId -eq $fldId_SubmitFrm_Quantity) {
                    $field | ConvertTo-JSON -Depth 10
                    $quantityValue = $field.valueAsNumber
}

Returned Value:

            {
              "TaskFormFieldId": 136139,
              "FieldId": 2092,
              "TaskFormSectionId": 19405,
              "DataTypeId": 3,
              "DataTypeName": "Whole Number",
              "Label": "Quantity",
              "ValueAsString": null,
              "ValueAsBoolean": null,
              "ValueAsNumber": null,
              "ValueAsDecimal": null,
              "ValueAsDate": null,
              "DefaultValue": null,
              "IsRequired": true,
              "IsVisible": true,
              "Position": 2,
              "Guid": 8003037754510,
              "ReferenceObject": null,
              "AppliedProperties": []
            }

**Comment by bmward80** [comment]
Hi, @IRodriguezSC,

Are you using `Get-Task` to retrieve the task for the before save method?

Brandon

**Reply by IRodriguezSC** [reply]
Yes, this is what I'm doing:

    $taskId = Get-InboundObjectId
    $task   = Get-Task $taskId -IncludeForms

    $task.Forms | ForEach-Object {
        
        $form = $_
        
        $form.Sections | ForEach-Object {
        
            $section = $_
    
            $section.Fields | ForEach-Object{
            
                $field = $_
                
                #--- Get office.
                if ($field.FieldId -eq $fldId_Office) {
                    $field | ConvertTo-JSON -Depth 10
                    $officeValue = $field.valueAsString
                }
    
                #--- Get quantity.
                if ($field.FieldId -eq $fldId_Quantity) {
                    $field | ConvertTo-JSON -Depth 10
                    $quantityValue = $field.valueAsNumber
                }
            }
        }
    }

**Reply by bmward80** [reply]
When you retrieve a task using `Get-Task`, you are retrieving the existing state of the task that's already saved in the system, not an incoming task change that is about to be saved. That is why you don't have the expected values, and why it works for an after save event. What you need to do is retrieve the incoming task that hasn't been saved yet from the `Get-InboundObjectInstance` cmdlet.

Here is some general information on before save triggers: [https://help.agilityblue.com/docs/scripting-tutorials-tasks-and-task-forms#tasks-attached-to-event-triggers](https://help.agilityblue.com/docs/scripting-event-triggers#before-save-action-state)

For tasks, there is some special considerations depending on the type of update action that you can read here: [https://help.agilityblue.com/docs/scripting-tutorials-tasks-and-task-forms#tasks-attached-to-event-triggers](https://help.agilityblue.com/docs/scripting-tutorials-tasks-and-task-forms#tasks-attached-to-event-triggers). This gives a nice example of how to approach task updates and how to work with the incoming data for a before save event.

Hopefully this gets you moving in the right direction,
Brandon

**Comment by IRodriguezSC** [comment, marked as answer]
Great thanks for the clarification, I will give that a try.

Edit: (removed all of your personal information)

**Reply by IRodriguezSC** [reply]
Using the Get-InboundObjectInstance cmdlet fixed the behavior.
